cluster/gce/windows/common.psm1

```diff
@@ -146,6 +146,320 @@ function MustDownload-File {
+  return $config
+}
+
+# Get-Backups returns a list of paths to backup files for the original file path -Path,
+# assuming that backup files are in the same directory, with a prefix matching
+# the original file name and a .zip suffix.
+function Get-Backups {
+  param (
+    # Original path of the file for which backups were created (no suffix).
+    [parameter(Mandatory=$true)] [string]$Path
+  )
+  $parent = Split-Path -Parent -Path $Path
+  $leaf = Split-Path -Leaf -Path $Path
+  $files = Get-ChildItem -File -Path $parent |
+           Where-Object Name -like "${leaf}*.zip"
+  return $files
+}
+
+# Trim-Backups deletes old backups for the log file identified by -Path until only -Count remain.
+# Deletes backups with the oldest CreationTime first.
+function Trim-Backups {
+  param (
+    [parameter(Mandatory=$true)] [int]$Count,
+    [parameter(Mandatory=$true)] [string]$Path
+  )
+  if ($Count -lt 0) {
+    $Count = 0
+  }
+  # If creating a new backup will exceed $Count, delete the oldest files
+  # until we have one less than $Count, leaving room for the new one.
+  # If the pipe results in zero items, $backups is $null, and if it results
+  # in only one item, PowerShell doesn't wrap in an array, so we check both cases.
+  # In the latter case, this actually caused it to often trim all backups, because
+  # .Length is also a property of FileInfo (size of the file)!
+  $backups = Get-Backups -Path $Path | Sort-Object -Property CreationTime
+  if ($backups -and $backups.GetType() -eq @().GetType() -and $backups.Length -gt $Count) {
+    $num = $backups.Length - $Count
+    $rmFiles = $backups | Select-Object -First $num
+    ForEach ($file in $rmFiles) {
+      Remove-Item $file.FullName
+    }
+  }
+}
+
+# Backup-File creates a compressed copy of the file at -Path.
+# The name of the backup is the same as the log file, with the suffix "-%Y%m%d-%s"
+# to identify the time of the backup.
+function Backup-File {
+  param (
+    [parameter(Mandatory=$true)] [string]$Path
+  )
+  $date = Get-Date -UFormat "%Y%m%d-%s"
+  $dest = "${Path}-${date}"
+  Copy-Item -Path $Path -Destination $dest
+  Compress-Archive -Path $dest -DestinationPath "${dest}.zip"
+  Remove-Item -Path $dest
+}
+
+# Rotate-File rotates the log file at -Path by first making a compressed copy of the original
+# log file with the suffix "-%Y%m%d-%s" to identify the time of the backup, then truncating
+# the original file in-place. Rotation is performed according to the options in -Config.
+function Rotate-File {
+  param (
+    # Path to the log file to rotate.
+    [parameter(Mandatory=$true)] [string]$Path,
+    # Config for file rotation.
+    [parameter(Mandatory=$true)] [FileRotationConfig]$Config
+  )
+  function rotate {
+    # If creating a new backup will exceed $MaxBackups, delete the oldest files
+    # until we have one less than $MaxBackups, leaving room for the new one.
+    Trim-Backups -Count ($Config.MaxBackups - 1) -Path $Path
+
+    Backup-File -Path $Path
```

Comment:
Hmm...can we do the operations in this order -- copy, truncate, and then compress?
Ideally we should minimize the time between copy & truncate to minimize log loss. Performing compression in the middle would probably increase the chance of log loss unnecessarily.

Reply:
Makes sense. Will do.